File: skills/runbook/scripts/commands/normalize.py
from __future__ import annotations

import argparse
import re
from pathlib import Path


NUMBERED_H3_RE = re.compile(r"^(\d+)\. (.+)$")
STEP_SIGNED_EXEC_RE = re.compile(r"^#### 执行 @\S+ \d{4}-\d{2}-\d{2} \d{2}:\d{2} [A-Za-z0-9:+-]+$")
STEP_SIGNED_ACCEPT_RE = re.compile(r"^#### 验收 @\S+ \d{4}-\d{2}-\d{2} \d{2}:\d{2} [A-Za-z0-9:+-]+$")
PLAN_JUMP_LINK_RE = re.compile(r"^\[跳转到(?:执行|验收)记录\]\(#item-\d+-(?:execution|acceptance)-record\)$")
ITEM_TOKEN_RE = re.compile(r"item-(\d+)(-execution-record|-acceptance-record)?")


def parse_sections(lines: list[str], level: int) -> list[tuple[int, str]]:
    prefix = "#" * level + " "
    return [
        (idx, line[len(prefix) :].strip())
        for idx, line in enumerate(lines)
        if line.startswith(prefix)
    ]


def section_slice(
    sections: list[tuple[int, str]], title: str, lines_len: int
) -> tuple[int, int] | None:
    for i, (start, name) in enumerate(sections):
        if name == title:
            end = sections[i + 1][0] if i + 1 < len(sections) else lines_len
            return start, end
    return None


def extract_h3_blocks(
    lines: list[str], start: int, end: int
) -> list[tuple[int, str, int, int]]:
    local_h3 = parse_sections(lines[start:end], 3)
    blocks: list[tuple[int, str, int, int]] = []
    for i, (local_start, title) in enumerate(local_h3):
        abs_start = start + local_start
        abs_end = start + local_h3[i + 1][0] if i + 1 < len(local_h3) else end
        blocks.append((abs_start, title, abs_start, abs_end))
    return blocks


def extract_h4_blocks(
    lines: list[str], start: int, end: int
) -> list[tuple[int, str, int, int]]:
    local_h4 = parse_sections(lines[start:end], 4)
    blocks: list[tuple[int, str, int, int]] = []
    for i, (local_start, title) in enumerate(local_h4):
        abs_start = start + local_start
        abs_end = start + local_h4[i + 1][0] if i + 1 < len(local_h4) else end
        blocks.append((abs_start, title, abs_start, abs_end))
    return blocks


def replace_item_number_tokens(line: str, old_number: int, new_number: int) -> str:
    def repl(match: re.Match[str]) -> str:
        actual = int(match.group(1))
        if actual != old_number:
            return match.group(0)
        suffix = match.group(2) or ""
        return f"item-{new_number}{suffix}"

    return ITEM_TOKEN_RE.sub(repl, line)


def normalize_numbered_step_section(
    lines: list[str],
    section_name: str,
    *,
    fix_item_tokens: bool,
) -> None:
    h2_sections = parse_sections(lines, 2)
    section = section_slice(h2_sections, section_name, len(lines))
    if section is None:
        return

    section_start, section_end = section
    expected = 1
    for heading_idx, title, block_start, block_end in extract_h3_blocks(lines, section_start + 1, section_end):
        match = NUMBERED_H3_RE.match(title)
        if match is None:
            continue

        actual = int(match.group(1))
        label = match.group(2)
        newline = "\n" if lines[heading_idx].endswith("\n") else ""
        lines[heading_idx] = f"### {expected}. {label}{newline}"

        if fix_item_tokens and actual != expected:
            anchor_idx = heading_idx - 1
            while anchor_idx > section_start and not lines[anchor_idx].strip():
                anchor_idx -= 1
            if anchor_idx > section_start:
                lines[anchor_idx] = replace_item_number_tokens(lines[anchor_idx], actual, expected)
            for idx in range(block_start + 1, block_end):
                lines[idx] = replace_item_number_tokens(lines[idx], actual, expected)
        expected += 1


def normalize_plan_jump_links(lines: list[str]) -> None:
    h2_sections = parse_sections(lines, 2)
    section = section_slice(h2_sections, "执行计划", len(lines))
    if section is None:
        return

    step_blocks = extract_h3_blocks(lines, section[0] + 1, section[1])
    for step_index, (_, _, step_start, step_end) in reversed(list(enumerate(step_blocks, start=1))):
        h4_blocks = extract_h4_blocks(lines, step_start + 1, step_end)
        for _, h4_title, block_start, block_end in reversed(h4_blocks):
            is_exec_block = h4_title == "执行" or STEP_SIGNED_EXEC_RE.match(f"#### {h4_title}") is not None
            is_accept_block = h4_title == "验收" or STEP_SIGNED_ACCEPT_RE.match(f"#### {h4_title}") is not None
            if not is_exec_block and not is_accept_block:
                continue

            expected_link = (
                f"[跳转到执行记录](#item-{step_index}-execution-record)"
                if is_exec_block
                else f"[跳转到验收记录](#item-{step_index}-acceptance-record)"
            )

            body_lines = [
                line
                for line in lines[block_start + 1 : block_end]
                if not PLAN_JUMP_LINK_RE.match(line.strip())
            ]
            while body_lines and not body_lines[0].strip():
                body_lines.pop(0)

            replacement = ["\n", f"{expected_link}\n", "\n", *body_lines]
            lines[block_start + 1 : block_end] = replacement


def normalize_runbook_numbering(text: str) -> str:
    lines = text.splitlines(keepends=True)
    if not lines:
        return text

    normalize_numbered_step_section(lines, "执行计划", fix_item_tokens=True)
    normalize_numbered_step_section(lines, "执行记录", fix_item_tokens=True)
    normalize_numbered_step_section(lines, "访谈记录", fix_item_tokens=False)
    normalize_plan_jump_links(lines)
    return "".join(lines)


def normalize_file(path: Path) -> tuple[str, str, bool]:
    source = path.read_text(encoding="utf-8")
    normalized = normalize_runbook_numbering(source)
    changed = normalized != source
    if changed:
        path.write_text(normalized, encoding="utf-8")
    return source, normalized, changed


def register(subparsers: argparse._SubParsersAction[argparse.ArgumentParser]) -> None:
    parser = subparsers.add_parser(
        "normalize",
        help="Normalize numbering, item anchors, and jump links",
        description="Normalize runbook numbering, item anchors, and jump links in place.",
    )
    parser.add_argument("path", help="Path to runbook markdown.")
    parser.set_defaults(handler=handle)


def handle(args: argparse.Namespace) -> int:
    path = Path(args.path).expanduser().resolve()
    _, _, changed = normalize_file(path)
    status = "updated" if changed else "unchanged"
    print(f"[runbook-normalize] {status} {path}")
    return 0

File: skills/runbook/scripts/commands/normalize.mjs
import path from "node:path";
import { promises as fs } from "node:fs";
import { splitLinesKeepEnds } from "./shared.mjs";

export const NUMBERED_H3_RE = /^(?:[🟢🟡🔴]\s+)?(\d+)\. (.+)$/u;
export const STEP_TRAFFIC_LIGHT_RE = /^([🟢🟡🔴]\s+)/u;
export const STEP_SIGNED_EXEC_RE = /^#### 执行 @\S+ \d{4}-\d{2}-\d{2} \d{2}:\d{2} [A-Za-z0-9:+-]+$/;
export const STEP_SIGNED_ACCEPT_RE = /^#### 验收 @\S+ \d{4}-\d{2}-\d{2} \d{2}:\d{2} [A-Za-z0-9:+-]+$/;
export const PLAN_JUMP_LINK_RE = /^\[跳转到(?:执行|验收)记录\]\(#item-\d+-(?:execution|acceptance)-record\)$/;
export const ITEM_TOKEN_RE = /item-(\d+)(-execution-record|-acceptance-record)?/g;

export function parseSections(lines, level) {
  const prefix = `${"#".repeat(level)} `;
  const result = [];
  for (let idx = 0; idx < lines.length; idx += 1) {
    if (lines[idx].startsWith(prefix)) {
      result.push([idx, lines[idx].slice(prefix.length).trim()]);
    }
  }
  return result;
}

export function sectionSlice(sections, title, linesLen) {
  for (let i = 0; i < sections.length; i += 1) {
    const [start, name] = sections[i];
    if (name === title) {
      const end = i + 1 < sections.length ? sections[i + 1][0] : linesLen;
      return [start, end];
    }
  }
  return null;
}

export function extractH3Blocks(lines, start, end) {
  const localH3 = parseSections(lines.slice(start, end), 3);
  return localH3.map(([localStart, title], index) => {
    const absStart = start + localStart;
    const absEnd = index + 1 < localH3.length ? start + localH3[index + 1][0] : end;
    return [absStart, title, absStart, absEnd];
  });
}

export function extractH4Blocks(lines, start, end) {
  const localH4 = parseSections(lines.slice(start, end), 4);
  return localH4.map(([localStart, title], index) => {
    const absStart = start + localStart;
    const absEnd = index + 1 < localH4.length ? start + localH4[index + 1][0] : end;
    return [absStart, title, absStart, absEnd];
  });
}

function replaceItemNumberTokens(line, oldNumber, newNumber) {
  return line.replace(ITEM_TOKEN_RE, (full, number, suffix = "") => {
    if (Number(number) !== oldNumber) {
      return full;
    }
    return `item-${newNumber}${suffix}`;
  });
}

function normalizeNumberedStepSection(lines, sectionName, { fixItemTokens }) {
  const h2Sections = parseSections(lines, 2);
  const section = sectionSlice(h2Sections, sectionName, lines.length);
  if (!section) {
    return new Map();
  }

  const [sectionStart, sectionEnd] = section;
  let expected = 1;
  const mapping = new Map();
  for (const [headingIdx, title, blockStart, blockEnd] of extractH3Blocks(lines, sectionStart + 1, sectionEnd)) {
    const match = title.match(NUMBERED_H3_RE);
    if (!match) {
      continue;
    }

    const actual = Number(match[1]);
    const label = match[2];
    mapping.set(actual, expected);
    const trafficLight = title.match(STEP_TRAFFIC_LIGHT_RE);
    const titlePrefix = trafficLight ? trafficLight[1] : "";
    const newline = lines[headingIdx].endsWith("\n") ? "\n" : "";
    lines[headingIdx] = `### ${titlePrefix}${expected}. ${label}${newline}`;

    if (fixItemTokens && actual !== expected) {
      let anchorIdx = headingIdx - 1;
      while (anchorIdx > sectionStart && !lines[anchorIdx].trim()) {
        anchorIdx -= 1;
      }
      if (anchorIdx > sectionStart) {
        lines[anchorIdx] = replaceItemNumberTokens(lines[anchorIdx], actual, expected);
      }
      for (let idx = blockStart + 1; idx < blockEnd; idx += 1) {
        lines[idx] = replaceItemNumberTokens(lines[idx], actual, expected);
      }
    }
    expected += 1;
  }
  return mapping;
}

function normalizeRollbackItemNumbers(lines, mapping) {
  if (!mapping.size) {
    return;
  }

  const h2Sections = parseSections(lines, 2);
  const section = sectionSlice(h2Sections, "回滚方案", lines.length);
  if (!section) {
    return;
  }

  const [start, end] = section;
  for (let idx = start + 1; idx < end; idx += 1) {
    const match = lines[idx].match(/^(\s*)(\d+)(\.\s+.*?)(\n?)$/);
    if (!match) {
      continue;
    }
    const oldNumber = Number(match[2]);
    const newNumber = mapping.get(oldNumber);
    if (!newNumber || newNumber === oldNumber) {
      continue;
    }
    lines[idx] = `${match[1]}${newNumber}${match[3]}${match[4]}`;
    lines[idx] = replaceItemNumberTokens(lines[idx], oldNumber, newNumber);
  }
}

function normalizePlanJumpLinks(lines) {
  const h2Sections = parseSections(lines, 2);
  const section = sectionSlice(h2Sections, "执行计划", lines.length);
  if (!section) {
    return;
  }

  const stepBlocks = extractH3Blocks(lines, section[0] + 1, section[1]);
  for (let i = stepBlocks.length - 1; i >= 0; i -= 1) {
    const stepIndex = i + 1;
    const [, , stepStart, stepEnd] = stepBlocks[i];
    const h4Blocks = extractH4Blocks(lines, stepStart + 1, stepEnd);
    for (let j = h4Blocks.length - 1; j >= 0; j -= 1) {
      const [, h4Title, blockStart, blockEnd] = h4Blocks[j];
      const isExecBlock = h4Title === "执行" || STEP_SIGNED_EXEC_RE.test(`#### ${h4Title}`);
      const isAcceptBlock = h4Title === "验收" || STEP_SIGNED_ACCEPT_RE.test(`#### ${h4Title}`);
      if (!isExecBlock && !isAcceptBlock) {
        continue;
      }

      const expectedLink = isExecBlock
        ? `[跳转到执行记录](#item-${stepIndex}-execution-record)`
        : `[跳转到验收记录](#item-${stepIndex}-acceptance-record)`;

      const bodyLines = lines
        .slice(blockStart + 1, blockEnd)
        .filter((line) => !PLAN_JUMP_LINK_RE.test(line.trim()));
      while (bodyLines.length > 0 && !bodyLines[0].trim()) {
        bodyLines.shift();
      }
      lines.splice(blockStart + 1, blockEnd - blockStart - 1, "\n", `${expectedLink}\n`, "\n", ...bodyLines);
    }
  }
}

export function normalizeRunbookNumbering(text) {
  const lines = splitLinesKeepEnds(text);
  if (lines.length === 0) {
    return text;
  }

  const planMapping = normalizeNumberedStepSection(lines, "执行计划", { fixItemTokens: true });
  normalizeRollbackItemNumbers(lines, planMapping);
  normalizeNumberedStepSection(lines, "执行记录", { fixItemTokens: true });
  normalizePlanJumpLinks(lines);
  return lines.join("");
}

export async function normalizeFile(filePath) {
  const source = await fs.readFile(filePath, "utf8");
  const normalized = normalizeRunbookNumbering(source);
  const changed = normalized !== source;
  if (changed) {
    await fs.writeFile(filePath, normalized, "utf8");
  }
  return { source, normalized, changed };
}

export async function handleNormalize(args) {
  const { changed } = await normalizeFile(path.resolve(args.path));
  const status = changed ? "updated" : "unchanged";
  console.log(`[runbook-normalize] ${status} ${path.resolve(args.path)}`);
  return 0;
}
